### PR TITLE
Improve azure determination method

### PR DESCRIPTION
**Pull request by sbernhard**
The "is likely running on azure is error prone":

The issue is mainly because of this two checks:
```
        if 'virt.host_type' in self.hw_info:
            if 'hyperv' in self.hw_info['virt.host_type']:
                probability += 0.3
```
```
        if found_microsoft is True:
            probability += 0.3
```
### Without this PR

**With python3-dmidecode**:
Azure: 0.9 (virt.host_type 0.3, chassis 0.3, found_microsoft 0.3)
HyperV: 0.6 (virt.host_type 0.3 and found_microsoft 0.3) => Issue false positive 0.6 => threshold 0.5 => is cloud/azure host would be true
This needs to be prevented.

**Without python3-dmidecode**:
Azure: 0.3 (virt.host_type 0.3) => issue. No azure detection.
HyperV: 0.3 (virt.host_type 0.3)

### Changes of this PR

This PR changes both probability calculation values from 0.3 to 0.2 (host_type) and 0.1 (microsoft). 
Additionally, it checks for waagent.log. The /var/log/waagent.log exists only on azure, but not on hyperV. The Azure Agent can be uninstalled, but with dmidecode installed, it would still be >= 0.5

### With PR 

**With python3-dmidecode**:
Azure: 1.0 (virt.host_type 0.2, chassis 0.3, found_microsoft 0.1, waagent 0.4). Without waagent: 0.6
HyperV: 0.3 (virt.host_type 0.2, found_microsoft 0.1)

**Without python3-dmidecode**:
Azure: 0.5 (virt.host_type 0.2, waagent 0.4)
HyperV: 0.2 (virt.host_type 0.2)

### CODE REVIEW

**Comment by m-horky** [comment]
Hi. The CI failed on two tests:

```
________ TestCloudProvider.test_detect_cloud_provider_azure_heuristics _________

self = <test_cloud_what.TestCloudProvider testMethod=test_detect_cloud_provider_azure_heuristics>
    def test_detect_cloud_provider_azure_heuristics(self):
        """
        Test the case, when detecting of azure does not work using strong signs, but it is necessary
        to use heuristics method
        """
        host_facts = {
            'virt.is_guest': True,
            'virt.host_type': 'hyperv',
            'dmi.bios.vendor': 'Microsoft',
            'dmi.bios.version': '1.0',
            'dmi.system.manufacturer': 'Google',
            'dmi.chassis.manufacturer': 'Amazon'
        }
        self.host_fact_collector_instance.get_all.return_value = host_facts
        detected_clouds = detect_cloud_provider()
>       self.assertEqual(detected_clouds, ['azure'])
E       AssertionError: Lists differ: [] != ['azure']
E       
E       Second list contains 1 additional elements.
E       First extra element 0:
E       'azure'
E       
E       - []
E       + ['azure']

test/cloud_what/test_cloud_what.py:1134: AssertionError
```

```
_______________ TestCloudProvider.test_conclict_in_strong_signs ________________

self = <test_cloud_what.TestCloudProvider testMethod=test_conclict_in_strong_signs>

    def test_conclict_in_strong_signs(self):
        """
        Test the case, when cloud providers change strong signs and there is conflict (two providers
        are detected using strong signs). In such case result using strong signs should be dropped
        and heuristics should be used, because strong signs do not work with probability and original
        order is influenced by the order of classes in 'constant' CLOUD_DETECTORS.
        """
        host_facts = {
            'virt.is_guest': True,
            'virt.host_type': 'kvm',
            'dmi.bios.vendor': 'Google',
            'dmi.bios.version': 'Amazon EC2',
            'dmi.chassis.asset_tag': '7783-7084-3265-9085-8269-3286-77',
            'dmi.chassis.manufacturer': 'Microsoft'
        }
        self.host_fact_collector_instance.get_all.return_value = host_facts
        detected_clouds = detect_cloud_provider()
        detected_clouds.sort()
>       self.assertEqual(detected_clouds, ['aws', 'azure', 'gcp'])
E       AssertionError: Lists differ: ['aws', 'gcp'] != ['aws', 'azure', 'gcp']
E       
E       First differing element 1:
E       'gcp'
E       'azure'
E       
E       Second list contains 1 additional elements.
E       First extra element 2:
E       'gcp'
E       
E       - ['aws', 'gcp']
E       + ['aws', 'azure', 'gcp']
E       ?         +++++++++

test/cloud_what/test_cloud_what.py:1154: AssertionError
```

Could you also include new test(s) for the new detection? E.g. one positive and one negative example, so the tests can show the tipping point more easily.

**Comment by sbernhard** [comment]
Thanks @m-horky 
I added some tests. Unfortunately, I need to adapt some tests, because the probability values need to be changed (otherwise a azure env would be recognized even if executed on a hyperV vm instance).

File: src/cloud_what/providers/azure.py
Comment (ptoscano) [inline src/cloud_what/providers/azure.py:117]: `E712 comparison to True should be 'if cond is True:' or 'if cond:'`

File: src/cloud_what/providers/azure.py
Comment (ptoscano) [inline src/cloud_what/providers/azure.py:159]: `E722 do not use bare 'except'`

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py]: Done. Thanks.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py]: Done. Thanks.

**Comment by m-horky** [comment]
Hi, thanks for the updates.

The PR seems fine to me now, but we'll wait for @jirihnidek's review before merging.

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:156]: I would like to know why did you decide to check for this file? Why don't you check for presence of `/usr/sbin/waagent`? Or even better why don't you use following construct:

```python
from shutil import which
return which('waagent') is not None
```

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:153]: I would like to see here some doc string.

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:116]: I'm not against the idea, but isn't it possible that some public/private cloud provider use this client too. It is open source and there is some chance that could be used outside of Azure. For this reason I would not increase the probability so much.

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:114]: I would rather increase probability in this case, because it is really specific for Azure ATM.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall I like this PR. I have some comments and concerns and you can find them in comments of code.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py:156]: Sure. This would be possible, too. There is no reason. Its just that I saw this log file.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py:153]: okay.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py:116]: Yes, but I don't think that some other cloud provider uses this agent. If it is the case, then we can adjust it later on, too.
The reason why I increased the "probability" so much is, that then the detection would also work if dmi-decode data doesn't exist.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py:114]: This chassis.asset_tag does only exist with dmi-decode data. If python-dmidecode isn't installed, this data doesn't exist. 

Additionally, the chassis.asset_tag is used in https://github.com/candlepin/subscription-manager/blob/0b7a210611de1dee494f3d29c8411dd3efa5d6b0/src/cloud_what/providers/azure.py#L89 as the single source to find out if sub-man is running on azure.

**Comment by barnabycourt** [comment]
ok to test

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [reply src/cloud_what/providers/azure.py:115]: The package `python-dmidecode` is required on x86_64 platform and I believe that Azure supports only this platform. Thus I believe that there is really low probability that python-dmidecode would not be installed.

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:132]: I don't like this change.

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [inline src/cloud_what/providers/azure.py:111]: I don't see the reason, why this probability should be decreased.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall, I like the idea that you added new fact that we can use for detection of Azure, but I don't like modification of probability for existing facts. I see risks of such changes. It is good to mention that final probability could be greater than `1.0`. 

If you have some serious issue with current code, then it would be fair to create Bugzilla bug report about it. It would help us understand the problem.

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py:111]: The reason why I changed the probability values of virt.host_type and found_microsoft:
- virt.host_type was at 0.3
- found_microsoft was at 0.3

On a pure hyperV, BOTH checks are true and therefore, hyperV is recognized as azure.Therefore I think, this need to be changed!

Comment (jirihnidek) [reply src/cloud_what/providers/azure.py:111]: I can see that `is_likely_running_on_cloud()` can give false positive result, but I think that it is good behavior. I will explain it later. First I would like to understand few things:

1) Did you observe this behavior, when subscription-manager was used? If yes, can you please describe this issue with more details?
2) Or do you use python-cloud-what as Python package and do you call `is_likely_running_on_cloud()` from your Python script/application? Why don't you use method `is_running_on_cloud()`?

Second I would like to clarify few things.

I would recommend to use method `is_running_on_cloud()` for detection if given public cloud is used. This should work, when it doesn't work, then it is big issue and we should fix it.

The method `is_likely_running_on_cloud()` was implemented as fallback, when `is_running_on_cloud()` returns false negative result. Thus it is good that `is_likely_running_on_cloud()` returns false positive result on Hyper-V. Without that we wouldn't be able to detect Azure, when Microsoft change strong sign. Anyway the method `is_likely_running_on_cloud()` should be used only as preliminary check if the VM is running on the given public cloud. When `is_likely_running_on_cloud()` returns some value greater than zero, then `get_metadata()` should be called. When this method returns some metadata, then only in this situation you have strong confidence that the VM is running on given public cloud.

BTW: the `provider.get_cloud_provider()` does all of this described in previous paragraph.

The method `detect_cloud_provider()` from `provider` module has default threshold=0.5, but it can be altered.

Comment (sbernhard) [reply src/cloud_what/providers/azure.py:111]: The error was observed by using the subscription-manager directly. 
Actually, if sub-man recognizes, that it runs on a azure (but actually it's a hyperV instance) it tries to read facts for azure, which fails:

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/subscription_manager/managercli.py", line 1863, in _do_command
    consumer = service.register(
  File "/usr/lib/python3/dist-packages/rhsmlib/services/register.py", line 85, in register
    facts_dict = self.facts.get_facts()
  File "/usr/lib/python3/dist-packages/subscription_manager/facts.py", line 78, in get_facts
    facts = collector.get_all()
  File "/usr/lib/python3/dist-packages/rhsmlib/facts/all.py", line 41, in get_all
    fact_collector = fact_collector_cls(collected_hw_info=results)
  File "/usr/lib/python3/dist-packages/rhsmlib/facts/cloud_facts.py", line 46, in __init__
    self.cloud_provider = get_cloud_provider(self._collected_hw_info)
  File "/usr/lib/python3/dist-packages/cloud_what/provider.py", line 149, in get_cloud_provider
    metadata = cloud_provider.get_metadata()
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 235, in get_metadata
    return super(AzureCloudProvider, self).get_metadata()
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 474, in get_metadata
    return self._get_metadata_from_server()
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 195, in _get_metadata_from_server
    metadata = super(AzureCloudProvider, self)._get_metadata_from_server()
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 402, in _get_metadata_from_server
    self._cached_metadata = self._get_data_from_server(
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 188, in _get_data_from_server
    return super(AzureCloudProvider, self)._get_data_from_server(data_type, url, headers)
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 387, in _get_data_from_server
    response = self._session.send(prepared_http_req, timeout=self.TIMEOUT)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='169.254.169.254', port=80): Read timed out. (read timeout=10.0)```

Therefore, it's pretty safe to detect azure by using the dmi.chassis.asset_tag and/or the availability of the azure linux agent. We should really prevent that sub-man has false-positive results. This is possible as described in this PR - even if dmidecode isn't available - which shouldn't be the case in 99,5% of the installations.

Comment (sbernhard) [reply src/cloud_what/providers/azure.py:111]: No. It's a private host with no access to the internet which often occurs in katello / rh satellite / orcharhino scenarios. 

Therefore, sub-man should not think its on a azure environment if the signs are pretty clear, that its not azure :)

Comment (jirihnidek) [reply src/cloud_what/providers/azure.py:111]: When you look at the traceback, then I can see that the exception `ReadTimeout` was raised. It means that there is some machine in your local network wit this IP and it is dropping all packets with HTTP/HTTPS requests. We should use more generic exception here:

https://github.com/candlepin/subscription-manager/blob/main/src/cloud_what/_base_provider.py#L386

Can you please try to replace `requests.ConnectionError` with `requests.RequestException`.

Comment (sbernhard) [reply src/cloud_what/providers/azure.py:111]: Well, actually I don't have access to this szenario anymore.  So, maybe this could be fixed but I think this would be another PR. 

As I said, in my eyes sub-man should never try to reach the azure metadata server if it's not a azure environment.

Comment (jirihnidek) [reply src/cloud_what/providers/azure.py:111]: Yeah, I will create another PR using `requests.RequestException`.

> As I said, in my eyes sub-man should never try to reach the azure metadata server if it's not a azure environment.

Only one fact (VM is running on Hyper-V) could not lead to the false positive detection of Azure. When hypervisor is Hyper-V, then this fact increase probability only to 0.3 and the threshold is 0.5. Thus there has to be another fact increasing the probability (e.g. `dmi.chassis.manufacturer` is `Microsoft`) In such case the total probability would be 0.6 and subscription-manager would try to contact IMDS server an get metadata.

From my perspective it is desired behavior and I can understand that you don't like it... So, there exists workaround for your case. You can create custom facts rewriting `dmi.chassis.manufacturer` with something like `Redacted` or `MS`:

```json
{
    "dmi.chassis.manufacturer": "Redacted"
}
```

Maybe it is also possible to modify information provided by SM BIOS from Hyper-V hypervisor.

Comment (sbernhard) [reply src/cloud_what/providers/azure.py:111]: > Only one fact (VM is running on Hyper-V) could not lead to the false positive detection of Azure.

Unfortunately, the facts look like this on a VM on pure HyperV:

```
{
  "dmi.chassis.manufacturer": "microsoft",
  "virt.host_type": "hyperv",
  "virt.is_guest": true,
}```

found_microsoft: +0.3
'hyperv' in self.hw_info['virt.host_type']: +0.3
--> 0.6 failure

**Comment by jirihnidek** [comment]
Created another PR fixing issue with printing of traceback to rhsm.log

https://github.com/candlepin/subscription-manager/pull/2984

File: src/cloud_what/providers/azure.py
Comment (jirihnidek) [reply src/cloud_what/providers/azure.py:111]: Well, I'm starting to consider using following approach: when we want to gather facts, then only detection using strong signs would be used and fallback to heuristics would be used only for automatic registration. We need false positive detection mostly due to automatic registration to be as reliable as possible. 

When detection using strong signs would not work (due to some unexpected change on metadata server of cloud provider) and not all facts would not be gathered/reported from metadata server, then we could fix this issue in strong detection and systems could be updated using dnf update.

Thus can you please change your PR and do not change existing probability?

**Comment by jirihnidek** [comment]
Created another BZ for this case: https://bugzilla.redhat.com/show_bug.cgi?id=2057053

File: src/cloud_what/providers/azure.py
Comment (sbernhard) [reply src/cloud_what/providers/azure.py]: Thank you very much. Understand your point now. 
I have adapted the PR again. Hope its now ready to be merged.

**Comment by sbernhard** [comment]
After reverting the change of the probability values, can this now be merged?

**Comment by sbernhard** [comment]
> After reverting the change of the probability values, can this now be merged @jirihnidek ?

**Comment by jirihnidek** [comment]
The code looks good. I'm testing it on all supported cloud providers...

**Review by jirihnidek — APPROVED** [review]
LGTM. Thanks for the PR and your contribution. :+1:

**Pull request merged**
